### PR TITLE
Support calculating number of chars for `StringArray`

### LINKED ISSUE #1493: Support calculate length by chars for `StringArray`
**Is your feature request related to a problem or challenge? Please describe what you are trying to do.**
Currently, we can only calculate the length of a string by byte. 
I guess it may be more user-friendly to provide a method that can calculate the length by char.

Re #1478.

### PR DESCRIPTION
**Pull request by HaoYang670**
# Which issue does this PR close?

Closes #1493 .

# Rationale for this change
 We need a method to calculate the number of chars for `StringArray`

# What changes are included in this PR?
Provide a pub method `num_chars`.

### CODE REVIEW

File: arrow/src/array/array_string.rs [hunk 1]
```diff
@@ -78,6 +78,39 @@ impl<OffsetSize: StringOffsetSizeTrait> GenericStringArray<OffsetSize> {
...
+    /// Returns the number of chars in the string at index `i`.
+    /// # Panic
+    /// If an invalid utf-8 byte is found, the function will panic.
+    /// However, this function does not check every byte. So you might
+    /// get an unexpected result if the string is in invalid utf-8 format.
+    /// # Performance
+    /// This function has `O(n)` time complexity where `n` is the string length.
+    /// If you can make sure that all chars in the string are in the range `U+0x0000` ~ `U+0x007F`,
+    /// please use the function [`value_length`](#method.value_length) which has O(1) time complexity.
+    pub fn num_chars(&self, i: usize) -> usize {
```
Comment (HaoYang670) [inline arrow/src/array/array_string.rs:90]: I don't find an elegant way to make the returned type as `OffsetSize`.

File: arrow/src/array/array_string.rs [hunk 2]
```diff
@@ -78,6 +78,39 @@ impl<OffsetSize: StringOffsetSizeTrait> GenericStringArray<OffsetSize> {
...
+        let offsets = self.value_offsets();
+        let start = offsets[i].to_usize().unwrap();
+        let end = offsets[i + 1].to_usize().unwrap();
+        let chars = &self.data.buffers()[1].as_slice()[start..end];
+
+        let mut char_iter = chars.iter();
+        let mut length: usize = 0;
+        while let Some(prefix) = char_iter.next() {
+            let ones = prefix.leading_ones() as usize;
+            match ones {
```
Comment (HaoYang670) [inline arrow/src/array/array_string.rs:100]: We just check the validation of the first byte for every char

File: arrow/src/array/array_string.rs [hunk 3]
```diff
@@ -78,6 +78,39 @@ impl<OffsetSize: StringOffsetSizeTrait> GenericStringArray<OffsetSize> {
...
 
+    /// Returns the number of chars in the string at index `i`.
+    /// # Panic
+    /// If an invalid utf-8 byte is found, the function will panic.
+    /// However, this function does not check every byte. So you might
+    /// get an unexpected result if the string is in invalid utf-8 format.
+    /// # Performance
+    /// This function has `O(n)` time complexity where `n` is the string length.
+    /// If you can make sure that all chars in the string are in the range `U+0x0000` ~ `U+0x007F`,
+    /// please use the function [`value_length`](#method.value_length) which has O(1) time complexity.
```
[the remark this answers is not in the document]
Comment (alamb) [reply arrow/src/array/array_string.rs:89]: I wonder if you have considered using `array.value(i).chars().len()` to count utf8 codepoints, as described in 
https://stackoverflow.com/questions/46290655/get-the-string-length-in-characters-in-rust?

File: arrow/src/array/array_string.rs (hunk 1 above)
Comment (alamb) [reply arrow/src/array/array_string.rs:90]: I think returning the length as a `usize` is a good API and is what would be expected by rust programmers 👍

File: arrow/src/array/array_string.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (HaoYang670) [reply arrow/src/array/array_string.rs:89]: > I wonder if you have considered using `array.value(i).chars().len()` to count utf8 codepoints, as described in https://stackoverflow.com/questions/46290655/get-the-string-length-in-characters-in-rust?

Thank you for your helpful suggestion, I will have a try!

[the remark this answers is not in the document]
Comment (HaoYang670) [reply arrow/src/array/array_string.rs:85]: Done!

File: arrow/src/array/array_string.rs [hunk 4]
```diff
@@ -377,9 +386,9 @@ mod tests {
 
     #[test]
     fn test_string_array_from_u8_slice() {
-        let values: Vec<&str> = vec!["hello", "", "parquet"];
+        let values: Vec<&str> = vec!["hello", "", "A£ऀ𖼚𝌆৩ƐZ"];
```
Comment (liukun4515) [inline arrow/src/array/array_string.rs:389]: 👍

**Review by liukun4515 — APPROVED** [review]
LGTM

**Comment by alamb** [comment]
Thanks @HaoYang670  and @liukun4515  for the review

**Pull request merged**